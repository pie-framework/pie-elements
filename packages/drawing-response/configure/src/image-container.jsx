import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';

import Button from './button';

const isImage = (file) => {
  const imageType = /image.*/;
  return file.type.match(imageType);
};

export class ImageContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      maxImageWidth: "100%",
      maxImageHeight: "auto",
      dragEnabled: true,
      dropzoneActive: false,
    };
  }

  componentDidMount() {
    if (this.imageSection) {
      const positionInfo = this.imageSection.getBoundingClientRect();
      const { height, width } = positionInfo;
      this.setState({
        maxImageWidth: width,
        maxImageHeight: height,
      });
    }
  }

  handleFileRead = (file) => {
    const { onImageUpload } = this.props;
    const reader = new FileReader();
    reader.onloadend = () => {
      onImageUpload(reader.result);
    };
    reader.readAsDataURL(file);
  };

  handleUploadImage = (e) => {
    e.preventDefault();
    const file = e.target.files[0];
    this.handleFileRead(file);
  };

  makeDropzoneActive = () => this.setState({ dropzoneActive: true });

  makeDropzoneInactive = () => this.setState({ dropzoneActive: false });

  handleOnDrop = (e) => {
    e.preventDefault();
    const { items, files } = e.dataTransfer;

    if (items && items[0].kind === 'file') {
      const file = items[0].getAsFile();
      if (isImage(file)) {
        this.handleFileRead(file);
      }
    } else if (isImage(files[0])) {
      this.handleFileRead(files[0]);
    }

    this.makeDropzoneInactive();
  };

  handleOnDragOver = (e) => {
    e.preventDefault();
    const { dropzoneActive } = this.state;

    if (!dropzoneActive) {
      this.makeDropzoneActive();
    }
  };

  handleOnDragExit = (e) => {
    e.preventDefault();
    this.makeDropzoneInactive();
  };

  handleEnableDrag = () => this.setState({ dragEnabled: true });
  handleDisableDrag = () => this.setState({ dragEnabled: false });
  handleInputClick = () => this.input.click();

  handleOnImageLoad = ({ target: { offsetHeight, offsetWidth, naturalHeight, naturalWidth } }) => {
    const { onUpdateImageDimension, imageDimensions } = this.props;
    const resizeHandle = this.resize;

    const dimensions = {
      height: (imageDimensions && imageDimensions.height) || offsetHeight || naturalHeight,
      width: (imageDimensions && imageDimensions.width) || offsetWidth || naturalWidth,
    };

    // check if aspect ratio is not respected on replacing image
    const imageAspectRatio = naturalWidth / naturalHeight;

    if (dimensions.width !== dimensions.height * imageAspectRatio) {
      dimensions.width = dimensions.height * imageAspectRatio;
    }

    this.setState({ dimensions });
    onUpdateImageDimension(dimensions);

    resizeHandle.addEventListener('mousedown', this.initialiseResize, false);
  };

  initialiseResize = () => {
    window.addEventListener('mousemove', this.startResizing, false);
    window.addEventListener('mouseup', this.stopResizing, false);
  };

  stopResizing = () => {
    const { onUpdateImageDimension } = this.props;
    const { dimensions } = this.state;

    this.handleEnableDrag();
    onUpdateImageDimension(dimensions);

    window.removeEventListener('mousemove', this.startResizing, false);
    window.removeEventListener('mouseup', this.stopResizing, false);
  };

  startResizing = (e) => {
    const box = this.image;
    const { maxImageWidth, maxImageHeight, dimensions } = this.state;

    const bounds = e.target.getBoundingClientRect();
    const x = e.clientX - bounds.left;
    const y = e.clientY - bounds.top;

    const imageAspectRatio = dimensions.width / dimensions.height;
    const fitsContainer = x <= maxImageWidth + 5 && (x / imageAspectRatio) <= maxImageHeight + 5;
    const hasMinimumWidth = x > 150 && y > 150;

    if (fitsContainer && hasMinimumWidth && box) {
      box.style.width = `${x}px`;
      box.style.height = `${x / imageAspectRatio}px`;

      this.setState({
        dimensions: {
          width: x,
          height: x / imageAspectRatio,
        },
      });
    }

    this.handleDisableDrag();
  };

  renderUploadControl(label) {
    const { classes } = this.props;

    return (
      <div>
        <Button label={label} onClick={this.handleInputClick} />
        <input
          accept="image/*"
          className={classes.input}
          onChange={this.handleUploadImage}
          ref={(ref) => {
            this.input = ref;
          }}
          type="file"
        />
      </div>
    );
  }

  render() {
    const { classes, imageUrl, imageDimensions } = this.props;
    const {
      dropzoneActive,
      dragEnabled,
      maxImageHeight,
      maxImageWidth
    } = this.state;

    return (
      <div className={classes.base}>
        <div
          className={`${classes.box} ${
            dropzoneActive ? classes.boxActive : ''
          }`}
          {...(dragEnabled
            ? {
                onDragExit: this.handleOnDragExit,
                onDragLeave: this.handleOnDragExit,
                onDragOver: this.handleOnDragOver,
                onDrop: this.handleOnDrop,
              }
            : {})}
        >
          <div className={classes.toolbar}>
            {this.renderUploadControl(
              imageUrl ? 'Replace Image' : 'Upload Image'
            )}
          </div>

          <div
            ref={(ref) => {
              this.imageSection = ref;
            }}
            className={classes.drawableHeight}
          >
            {imageUrl ? (
              <div className={classes.imageContainer}>
                <img
                  className={classes.image}
                  height="auto"
                  onLoad={this.handleOnImageLoad}
                  ref={(ref) => {
                    this.image = ref;
                  }}
                  src={imageUrl}
                  style={{
                    width: imageDimensions && imageDimensions.width ? imageDimensions.width : undefined,
                    maxWidth: maxImageWidth,
                    maxHeight: maxImageHeight
                  }}
                  alt=""
                />
                <div
                  ref={(ref) => {
                    this.resize = ref;
                  }}
                  className={classes.resize}
                />
              </div>
            ) : (
              <div className={`${classes.drawableHeight} ${classes.centered}`}>
                <label>Drag and drop or upload image from computer</label>
                <br />
                {this.renderUploadControl('Upload Image')}
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}

const styles = (theme) => ({
  base: {
    marginTop: theme.spacing.unit * 2,
  },
  box: {
    border: '1px solid #E0E1E6',
    borderRadius: '5px',
  },
  boxActive: {
    border: '1px solid #0032C2',
  },
  centered: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
  },
  drawableHeight: {
    minHeight: 350,
  },
  image: {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'center',
  },
  imageContainer: {
    position: 'relative',
    width: 'fit-content',
  },
  resize: {
    borderBottom: '1px solid #727272',
    borderRight: '1px solid #727272',
    bottom: '-10px',
    cursor: 'se-resize',
    height: '10px',
    position: 'absolute',
    right: '-10px',
    width: '10px',
  },
  input: {
    display: 'none',
  },
  toolbar: {
    backgroundColor: '#ECEDF1',
    borderBottom: '1px solid #E0E1E6',
    borderTopLeftRadius: '5px',
    borderTopRightRadius: '5px',
    display: 'flex',
    padding: '12px 8px',
  },
});

ImageContainer.propTypes = {
  classes: PropTypes.object.isRequired,
  imageUrl: PropTypes.string.isRequired,
  onImageUpload: PropTypes.func.isRequired,
  onUpdateImageDimension: PropTypes.func.isRequired,
};

export default withStyles(styles)(ImageContainer);
