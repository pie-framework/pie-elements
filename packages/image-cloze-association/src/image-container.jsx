import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core';

import ImageDropTarget from './image-drop-target';

class ImageContainer extends Component {
  render() {
    const {
      answers,
      canDrag,
      classes,
      draggingElement,
      image: { height, src, width } = {},
      onAnswerSelect,
      onDragAnswerBegin,
      onDragAnswerEnd,
      responseContainers,
      showDashedBorder,
      responseAreaFill,
      answerChoiceTransparency,
      responseContainerPadding,
      imageDropTargetPadding,
    } = this.props;

    return (
      <div className={classes.base}>
        <img src={src} height={height} width={width} alt={'Image not found'} />

        {(responseContainers || []).map((r, i) => {
          const rHeight = (r.height.replace('%', '') / 100) * height;
          const rWidth = (r.width.replace('%', '') / 100) * width;
          const rLeft = (r.x / 100) * width;
          const rTop = (r.y / 100) * height;
          const answersParsed = answers.filter((a) => a.containerIndex === r.index);

          return (
            <ImageDropTarget
              answers={answersParsed}
              canDrag={canDrag}
              containerStyle={{
                height: rHeight,
                width: rWidth,
                left: rLeft,
                top: rTop,
              }}
              key={r.id + i}
              draggingElement={draggingElement}
              onDrop={(item) => onAnswerSelect(item, r.index)}
              onDragAnswerBegin={onDragAnswerBegin}
              onDragAnswerEnd={onDragAnswerEnd}
              showDashedBorder={showDashedBorder}
              responseAreaFill={responseAreaFill}
              answerChoiceTransparency={answerChoiceTransparency}
              responseContainerPadding={responseContainerPadding}
              imageDropTargetPadding={imageDropTargetPadding}
            />
          );
        })}
      </div>
    );
  }
}

ImageContainer.propTypes = {
  answers: PropTypes.array.isRequired,
  canDrag: PropTypes.bool.isRequired,
  classes: PropTypes.object,
  draggingElement: PropTypes.shape({}).isRequired,
  image: PropTypes.object.isRequired,
  onAnswerSelect: PropTypes.func.isRequired,
  onDragAnswerBegin: PropTypes.func.isRequired,
  onDragAnswerEnd: PropTypes.func.isRequired,
  responseContainers: PropTypes.array.isRequired,
  showDashedBorder: PropTypes.bool,
  answerChoiceTransparency: PropTypes.bool,
  responseAreaFill: PropTypes.string,
  responseContainerPadding: PropTypes.string,
  imageDropTargetPadding: PropTypes.string,
};

ImageContainer.defaultProps = {
  classes: {},
};

const styles = (theme) => ({
  base: {
    margin: theme.spacing.unit * 2,
    position: 'relative',
    width: 'fit-content',
  },
});

export default withStyles(styles)(ImageContainer);
