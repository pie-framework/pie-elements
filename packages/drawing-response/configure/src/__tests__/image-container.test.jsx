import { shallow } from 'enzyme';
import React from 'react';

import { ImageContainer } from '../image-container';

jest.mock('@pie-lib/config-ui', () => ({
  settings: {
    Panel: props => <div {...props} />,
    toggle: jest.fn(),
    radio: jest.fn(),
    dropdown: jest.fn()
  },
  layout: {
    ConfigLayout: props => <div>{props.children}</div>
  }
}));

describe('ImageContainer', () => {
  const onUpdateImageDimension = jest.fn();
  const onImageUpload = jest.fn();
  let wrapper;

  beforeEach(() => {
    const props = {
      classes: {},
      imageUrl: 'url',
      onUpdateImageDimension,
      onImageUpload
    };
    wrapper = () => shallow(<ImageContainer {...props} />);
  });

  describe('snapshot', () => {
    it('renders', () => {
      expect(wrapper()).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    let w;
    let oldFileReader;

    beforeAll(() => {
      const dummyFileReader = {};

      w = wrapper();
      w.instance().resize = { addEventListener: jest.fn() };

      oldFileReader = window.FileReader;

      dummyFileReader.readAsDataURL = (result) => {
        dummyFileReader.result = result;
        dummyFileReader.onloadend();
      };
      window.FileReader = jest.fn(() => dummyFileReader);
    });

    it('handleFileRead calls onImageUpload', () => {
      w.instance().handleFileRead('file');

      expect(onImageUpload).toBeCalled();
      expect(onImageUpload).toHaveBeenCalledTimes(1);
    });

    it('handleUploadImage calls handleFileRead which calls onImageUpload', () => {
      w.instance().handleUploadImage({ target: { files: ['file'] }, preventDefault: jest.fn() });

      expect(onImageUpload).toBeCalled();
      expect(onImageUpload).toHaveBeenCalledTimes(2);
    });

    it('handleOnDrop calls handleFileRead which calls onImageUpload if item is file and image', () => {
      w.instance().handleOnDrop({
        preventDefault: jest.fn(),
        dataTransfer: {
          items: [{
            kind: 'file',
            getAsFile: jest.fn().mockReturnValue({ type: 'image', key: 'item' })
          }
          ],
          files: []
        }
      });

      expect(onImageUpload).toBeCalledWith({ type: 'image', key: 'item' });
      expect(onImageUpload).toHaveBeenCalledTimes(3);
    });

    it('handleOnDrop calls handleFileRead which doesn\'t call onImageUpload if item is file but not image', () => {
      w.instance().handleOnDrop({
        preventDefault: jest.fn(),
        dataTransfer: {
          items: [{
            kind: 'file',
            getAsFile: jest.fn().mockReturnValue({ type: 'jpg' })
          }
          ],
          files: []
        }
      });

      // same times number as in previous test, meaning that is was not called again
      expect(onImageUpload).toHaveBeenCalledTimes(3);
    });

    it('handleOnDrop calls handleFileRead which calls onImageUpload if item is not file, but file is image', () => {
      w.instance().handleOnDrop({
        preventDefault: jest.fn(),
        dataTransfer: {
          items: [{
            kind: 'something else',
            getAsFile: jest.fn().mockReturnValue({ type: 'image' })
          }
          ],
          files: [{ type: 'image', key: 'file' }]
        }
      });

      expect(onImageUpload).toBeCalledWith({ type: 'image', key: 'file' });
      expect(onImageUpload).toHaveBeenCalledTimes(4);
    });

    it('handleOnDrop calls handleFileRead which doesn\'t call onImageUpload if item is not file and file is not image', () => {
      w.instance().handleOnDrop({
        preventDefault: jest.fn(),
        dataTransfer: {
          items: [{
            kind: 'something else',
            getAsFile: jest.fn().mockReturnValue({ type: 'image' })
          }
          ],
          files: [{ type: 'something else', key: 'file' }]
        }
      });

      // same times number as in previous test, meaning that is was not called again
      expect(onImageUpload).toHaveBeenCalledTimes(4);
    });

    it('handleOnImageLoad calls onUpdateImageDimension', () => {
      w.instance().handleOnImageLoad({ target: { offsetHeight: 50, offsetWidth: 50, naturalWidth: 50, naturalHeight: 50 }});

      expect(onUpdateImageDimension).toHaveBeenCalledWith({
        height: 50,
        width: 50
      });
    });

    it('stopResizing calls onUpdateImageDimension', () => {
      w.instance().handleOnImageLoad({ target: { offsetHeight: 50, offsetWidth: 50, naturalWidth: 50, naturalHeight: 50 }});

      expect(onUpdateImageDimension).toHaveBeenCalledWith(w.instance().state.dimensions);
    });

    afterAll(() => {
      window.FileReader = oldFileReader;
    });
  });
});

