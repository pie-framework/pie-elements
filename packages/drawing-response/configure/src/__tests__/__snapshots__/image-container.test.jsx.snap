// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageContainer snapshot renders 1`] = `
<div>
  <div
    className="undefined "
    onDragExit={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <div>
      <div>
        <WithStyles(RawButton)
          label="Replace Image"
          onClick={[Function]}
        />
        <input
          accept="image/*"
          onChange={[Function]}
          type="file"
        />
      </div>
    </div>
    <div>
      <div>
        <img
          alt=""
          height="auto"
          onLoad={[Function]}
          src="url"
          style={
            Object {
              "maxHeight": "auto",
              "maxWidth": "100%",
              "width": undefined,
            }
          }
        />
        <div />
      </div>
    </div>
  </div>
</div>
`;
